aes/aesni/src/aes192/expand.rs

```diff
@@ -1,91 +1,107 @@
+        t4 = _mm_slli_si128(t3, 0x4);
+        t3 = _mm_xor_si128(t3, t4);
+        t3 = _mm_xor_si128(t3, t2);
+
+        (t1, t3)
+    }}
 }
 
 macro_rules! shuffle {
     ($a:expr, $b:expr, $imm:expr) => {
-        unsafe {
-            mem::transmute::<_, __m128i>(
-                _mm_shuffle_pd(mem::transmute($a), mem::transmute($b), $imm)
-            )
-        }
+        mem::transmute::<_, __m128i>(
+            _mm_shuffle_pd(mem::transmute($a), mem::transmute($b), $imm)
+        )
     }
 }
 
 #[inline(always)]
 pub(super) fn expand(key: &[u8; 24]) -> ([__m128i; 13], [__m128i; 13]) {
-    let mut enc_keys: [__m128i; 13] = unsafe { mem::uninitialized() };
-    let mut dec_keys: [__m128i; 13] = unsafe { mem::uninitialized() };
-
-    let (t1, t3) = unsafe {
-        (
-            _mm_loadu_si128(key.as_ptr() as *const __m128i),
-            // we copy garbage for second half, it will be overwritten, so it's fine
-            _mm_loadu_si128(key.as_ptr().offset(16) as *const __m128i),
-        )
-    };
-    enc_keys[0] = t1;
-    enc_keys[1] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x01);
-    enc_keys[1] = shuffle!(enc_keys[1], t1, 0);
-    enc_keys[2] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x02);
-    enc_keys[3] = t1;
-    enc_keys[4] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x04);
-    enc_keys[4] = shuffle!(enc_keys[4], t1, 0);
-    enc_keys[5] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x08);
-    enc_keys[6] = t1;
-    enc_keys[7] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x10);
-    enc_keys[7] = shuffle!(enc_keys[7], t1, 0);
-    enc_keys[8] = shuffle!(t1, t3, 1);
-    let (t1, t3) = expand_round!(t1, t3, 0x20);
-    enc_keys[9] = t1;
-    enc_keys[10] = t3;
-
-    let (t1, t3) = expand_round!(t1, t3, 0x40);
-    enc_keys[10] = shuffle!(enc_keys[10], t1, 0);
-    enc_keys[11] = shuffle!(t1, t3, 1);
-    let (t1, _) = expand_round!(t1, t3, 0x80);
-    enc_keys[12] = t1;
-
-    dec_keys[0] = enc_keys[0];
-    for i in 1..12 {
-        dec_keys[i] = unsafe { _mm_aesimc_si128(enc_keys[i]) };
-    }
-    dec_keys[12] = enc_keys[12];
+    unsafe {
+        let mut enc_keys: [__m128i; 13] = mem::uninitialized();
+        let mut dec_keys: [__m128i; 13] = mem::uninitialized();
 
-    (enc_keys, dec_keys)
+        macro_rules! store {
+            ($i:expr, $k:expr) => {
+                _mm_store_si128(enc_keys.as_mut_ptr().offset($i), $k);
+                _mm_store_si128(dec_keys.as_mut_ptr().offset($i),
+                    _mm_aesimc_si128($k));
+            }
+        }
+
+        // we are being extra pedantic here to remove out-of-bound access.
+        // this should be optimized out into movups, movsd sequence
+        // note that unaligned load MUST be used here, even though we read
+        // from the array (compiler missoptimizes aligned load)
+        let (k0, k1l) = {
+            let mut t: [u8; 32] = mem::uninitialized();
```

Comment:
You can use `#[repr(align(16))]` here to force this array to be properly aligned, so that you can use aligned loads and stores. Something like:

```rust
#[repr(align(16))] struct u8x32([u8; 32]);
```

would do.

Reply:
Either way we'll need unaligned load here, because `key` pointer can be misaligned. I've commented it because for some strange reason I get `SIGSEGV: invalid memory reference` when I replace `_mm_loadu_si128` with `_mm_load_si128`, even though godbolt snippet (linked in the issue) results in movups for both variants.

Reply:
Hm. `key` is just loaded by `*key`, and written to the first 24 bytes of the `[u8; 32]`. 

There are, however, two `_mm_loadu_si128` below that read from the `[u8; 32]` on the stack. These could become aligned, if the `[u8; 32]` were aligned to a 16 byte boundary. Using `#[repr(align(16))]` would allow that.

Whether this influences code generation or not, no idea. Hopefully LLVM optimizes all of this to two vector loads from `key` in which case these vector loads need to be unaligned loads.

Reply:
I think there is a bug somewhere in the LLVM codegen, but I can't produce a minimal example to demonstrate it... Even if I'll give the whole `expand` function to godbolt it produces the same assembly for both `_mm_loadu_si128` and `_mm_load_si128`...

Reply:
Yes it does not create temporary, as was demonstrated in the link posted in the issue and can be seen [here](https://godbolt.org/g/z4PQtN).